packages/core/src/protocol/Invoice.ts

```diff
@@ -0,0 +1,335 @@
+import { z } from "zod";
+import { bech32, bech32m } from "bech32";
+import { KycStatus, kycStatusFromBytes, kycStatusToBytes } from "./KycStatus.js";
+import {
+    deserializeBoolean,
+    deserializeNumber,
+    deserializeString,
+    serializeBoolean,
+    serializeNumber,
+    serializeString,
+} from "../serializerUtils.js";
+import { optionalIgnoringNull } from "../zodUtils.js";
+import {
+    CounterPartyDataOptionsSchema,
+    counterPartyDataOptionsFromBytes,
+    counterPartyDataOptionsToBytes
+} from "./CounterPartyData.js";
+
+
+const UMA_BECH32_PREFIX = "uma";
+const BECH_32_MAX_LENGTH = 512;
+
+const InvoiceCurrencySchema = z.object({
+    name: z.string(),
+    code: z.string(),
+    symbol: z.string(),
+    decimals: z.number()
+})
+
+/**
+ * Sub object describing receiving currency in more detail
+ * 
+ * @param name
+ * @param code
+ * @param symbol
+ * @param decimals
+ */
+export type InvoiceCurrency = z.infer<typeof InvoiceCurrencySchema>;
+
+
+const InvoiceSchema = z.object({
+    receiverUma: z.string(),
+    invoiceUUID: z.string().uuid(),
+    amount: z.number(),
+    receivingCurrency: InvoiceCurrencySchema,
+    expiration: z.number(),
+    isSubjectToTravelRule: z.boolean(),
+    requiredPayerData: optionalIgnoringNull(CounterPartyDataOptionsSchema),
+    umaVersion: z.string(),
+    commentCharsAllowed: optionalIgnoringNull(z.number()),
+    senderUma: optionalIgnoringNull(z.string()),
+    invoiceLimit: optionalIgnoringNull(z.number()),
+    kycStatus: optionalIgnoringNull(z.nativeEnum(KycStatus)),
+    callback: z.string(),
+    signature: optionalIgnoringNull(z.instanceof(Uint8Array))
```

Comment:
@zhenlu I hadn't really understood this field until looking at the vasp sample app - 
signature is the signed, tlv encoded Invoice , so the correct creation flow is 
```
create invoice = Invoice (w/o signature set)
signature = sign(signing key, invoice.tlv())
invoice.signature = signature
```
is that correct? if so I'll update how this is created